// views/m/common/CitySelector/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as Vue from 'vue';
import * as autocomplete from 'autocomplete';

class VPresenter extends project.VPresenter {
    private _fromPage: string;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._fromPage = 'Index';
        var a = autocomplete;//hack
        this._watchEvent();
        this._els = this._getElements();
        let source:any[] = JSON.parse(this._els['source'][0]['text']);
        let cityList:{label:string,value:string}[] = [];
        source.forEach(function(item){
            let list:string[] = item.list;
            list.forEach(function(city){
                cityList.push({label:city,value:city});
            });
        });     
        var v_currCity = this.find('#CitySelector-current')[0];
        var curr_city = model.globalData.city;
        new Vue({
            el: v_currCity,
            data: curr_city
        })       
        //搜索城市自动完成提示
        this.find('.input-div').autocomplete({
            source: cityList,
            //选中事件
            select: (event, ui)=>{
                $(event.target).val(ui.item.label);                
                event.preventDefault();
                this.select(ui.item.label,ui.item.value);
            }
        }).autocomplete('instance')._renderItem = function (ul, item) {
            return $("<li><a href='#'>" + item.label + "</a></li>").appendTo(ul);
        }        
    }
    /**repair */
    public  setProv(formPage: string) {
        this._fromPage = formPage;
    }
    _evt_selected(data:any,target:HTMLElement,hit:HTMLElement){
        let city = hit.getAttribute("data-city");
        if(city){
            this.select(city,city);
        }
    }
    select(text:string,id:string){
        model.globalData.selected_city.city_id = id;
        model.globalData.selected_city.city_name = text;
        model.globalData.selected_province.province_id = model.globalData.cityToProvince[id] || id;
        model.globalData.selected_province.province_name = model.globalData.cityToProvince[id] || id;
        if (this._fromPage == 'Index') {
            tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/Index'), (mod:tomato.VPresenter) => {
                funs.loadPage(mod);
            });
        }else if (this._fromPage == 'Repair') {
            tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/Repair'), (mod:tomato.VPresenter) => {
                funs.loadPage(mod);
            });
        }else if(this._fromPage=="select"){
            this._evt_close();
        }
        
    }
}

export = VPresenter;
