// nodejs/global/api/api.ts
import * as potato from '@po-to/potato';
import * as user from 'nodejs/global/model/user';
import * as city from 'nodejs/global/model/city';
import * as init from 'nodejs/global/model/init';
import * as Demand from 'nodejs/global/model/demand';

let emptyFailHandler = function () { }
let apiServer = "`{APIURL}`";
//let apiServer = "http://s.m.88ba.com";
let ApiDomain = apiServer.substr(apiServer.indexOf("//") + 2);

function budget(money: number): { budget: string, unit: string } {
    function std_money_format_in_th(money) {
        money = money + "";
        var intLen = money.split('.')[0].length;
        if (intLen > 4) {
            return { "budget": std_num_format(money / 10000), "unit": "万" }
        } else if (intLen > 3) {
            return { "budget": std_num_format(money / 1000), "unit": "千" }
        } else if (intLen > 2) {
            return { "budget": std_num_format(money / 100), "unit": "百" }
        } else {
            return { "budget": money, "unit": "" };
        }
    }
    function std_num_format(num) {
        num += '';
        var parts = num.split('.');
        var float_cnt = 0;
        if (parts.length > 1) {
            var deci = parts[1];
            float_cnt = (parseInt(deci) > 0) ? deci.length : 0;
        }

        var pureRes = number_format(num, Math.min(2, float_cnt));
        var ptIndex = pureRes.indexOf('.');
        if (ptIndex === -1) {
            // 整数
            return pureRes;
        } else {
            // 下面这些处理，是为了死也要保证小数点后末尾不能为0。即使这是不科学的四舍五入，产品及相关人员也认为这比科学要更好。好吧，我妥协
            var num: any = parseInt(pureRes.substr(ptIndex + 1));
            if (0 === num) {
                // 123.00型
                return pureRes.substr(0, ptIndex);
            } else if (0 === num % 10) {
                // 123.40型
                return pureRes.substr(0, pureRes.length - 1);
            } else {
                return pureRes;
            }
        }
    }
    function number_format(number, decimals, decPoint?, thousandsSep?) { // eslint-disable-line camelcase

        number = (number + '').replace(/[^0-9+\-Ee.]/g, '')
        var n = !isFinite(+number) ? 0 : +number
        var prec = !isFinite(+decimals) ? 0 : Math.abs(decimals)
        var sep = (typeof thousandsSep === 'undefined') ? ',' : thousandsSep
        var dec = (typeof decPoint === 'undefined') ? '.' : decPoint
        var s: any = '';
        var toFixedFix = function (n, prec) {
            var k = Math.pow(10, prec)
            return '' + (Math.round(n * k) / k)
                .toFixed(prec)
        }
        // @todo: for IE parseFloat(0.55).toFixed(0) = 0;
        s = (prec ? toFixedFix(n, prec) : '' + Math.round(n)).split('.')
        if (s[0].length > 3) {
            s[0] = s[0].replace(/\B(?=(?:\d{3})+(?!\d))/g, sep)
        }
        if ((s[1] || '').length < prec) {
            s[1] = s[1] || ''
            s[1] += new Array(prec - s[1].length + 1).join('0')
        }

        return s.join(dec)
    }
    let data = std_money_format_in_th(money);
    if (data.budget == "" || data.budget == 0) {
        data.budget = "";
    }
    return data;
    // if (money.budget == "" || money.budget == 0) {
    //     money.budget = "议价";
    // }
    // return money;
    // this.find('.format-budget').html(money.budget).next('.budget-unit').html(money.unit);
}
let demand_empty = {
    "user": {
        id: '',
        nick: '',
        sex: 0,
    },
    "id": "", //项目ID；
    "cate": '',
    "cate_name": '',
    "city_id": "",
    "city_name": "",
    "malf":'',
    "malf_name":"",
    "prov": "",
    "address": "",
    "type": 2,	//分类：0：其它；1：门头招牌；2：户外广告牌；3：信息告示牌；4：舞台用屏；5：室内高清屏；
    "size": "",
    "location": 1,  //1 => '户外', 2 => '室内', 3 => '半户外',
    "color": 1, //1：单色，2：双色，3：全彩
    "span": "",
    "budget": 0, //预算，0：表示议价；
    "budgetFormat": budget(0),
    "image": "",	//需求场地图片
    "status": "", //10进行中，60：交易结束
    "intm": 0,
    "uptm": 0,
    "note": "",
    "audio": "", //客服和客户之间的录音链接
    "bid_fee": 0, //抢单服务费
    "type_name": "户外广告屏",
    "location_name": '户外',
    "color_name": "单色",
    "span_id": 1,
    "supplier_list": []
}
let demand_config = {

    typeOptions: {
        1: {
            name: '门头字幕屏', id: '1', def: {
                location: '3', color: '1', span: '3',
                locationOptions: { 1: { name: '户外', id: '1' }, 2: { name: '室内', id: '2' }, 3: { name: '半户外', id: '3' } },
                colorOptions: { 1: { name: '单色', id: '1' }, 2: { name: '双色', id: '2' }, 3: { name: '全彩', id: '3' } },
                spanOptions: { 1: { name: 'P6', id: '1' }, 2: { name: 'P8', id: '2' }, 3: { name: 'P10', id: '3' }, 4: { name: '其他', id: '4' }, },
            }
        },
        2: {
            name: '户外广告屏', id: '2', def: {
                location: '1', color: '3', span: '3',
                locationOptions: { 1: { name: '户外', id: '1' } },
                colorOptions: { 1: { name: '单色', id: '1' }, 2: { name: '双色', id: '2' }, 3: { name: '全彩', id: '3' } },
                spanOptions: { 1: { name: 'P6', id: '1' }, 2: { name: 'P8', id: '2' }, 3: { name: 'P10', id: '3' }, 4: { name: '其他', id: '4' }, },
            }
        },
        3: {
            name: '信息告示屏', id: '3', def: {
                location: '1', color: '1', span: '3',
                locationOptions: { 1: { name: '户外', id: '1' }, 2: { name: '室内', id: '2' }, 3: { name: '半户外', id: '3' } },
                colorOptions: { 1: { name: '单色', id: '1' }, 2: { name: '双色', id: '2' }, 3: { name: '全彩', id: '3' } },
                spanOptions: { 1: { name: 'P6', id: '1' }, 2: { name: 'P8', id: '2' }, 3: { name: 'P10', id: '3' }, 4: { name: '其他', id: '4' }, },
            }
        },
        4: {
            name: '舞台用屏', id: '4', def: {
                location: '2', color: '3', span: '2',
                locationOptions: { 1: { name: '户外', id: '1' }, 2: { name: '室内', id: '2' }, 3: { name: '半户外', id: '3' } },
                colorOptions: { 1: { name: '单色', id: '1' }, 2: { name: '双色', id: '2' }, 3: { name: '全彩', id: '3' } },
                spanOptions: { 1: { name: 'P3', id: '1' }, 2: { name: 'P4', id: '2' }, 3: { name: 'P5', id: '3' }, 4: { name: '其他', id: '4' } }
            }
        },
        5: {
            name: '室内高清屏', id: '5', def: {
                location: '2', color: '3', span: '2',
                locationOptions: { 2: { name: '室内', id: '2' } },
                colorOptions: { 1: { name: '单色', id: '1' }, 2: { name: '双色', id: '2' }, 3: { name: '全彩', id: '3' } },
                spanOptions: { 1: { name: 'P1.25', id: '1' }, 2: { name: 'P2.0', id: '2' }, 3: { name: 'P2.5', id: '3' }, 4: { name: '其他', id: '4' } }
            }
        },
        0: {
            name: '其他', id: '0', def: {
                location: '1', color: '3', span: '3',
                locationOptions: { 1: { name: '户外', id: '1' }, 2: { name: '室内', id: '2' }, 3: { name: '半户外', id: '3' } },
                colorOptions: { 1: { name: '单色', id: '1' }, 2: { name: '双色', id: '2' }, 3: { name: '全彩', id: '3' } },
                spanOptions: { 1: { name: 'P6', id: '1' }, 2: { name: 'P8', id: '2' }, 3: { name: 'P10', id: '3' }, 4: { name: '其他', id: '4' } },
            }
        }
    },
    locationOptions: {
        1: { name: '户外', id: '1' },
        2: { name: '室内', id: '2' },
        3: { name: '半户外', id: '3' }
    },
    colorOptions: {
        1: { name: '单色', id: '1' },
        2: { name: '双色', id: '2' },
        3: { name: '全彩', id: '3' }
    },
    spanOptions: {
        1: { name: 'P6', id: '1' },
        2: { name: 'P8', id: '2' },
        3: { name: 'P10', id: '3' },
        4: { name: '其他', id: '4' },
    },
    faultOptions: {
        1: { name: '整屏或部分不亮', id: '1' },
        2: { name: '出现亮点暗点', id: '2' },
        3: { name: '有色彩问题', id: '3' },
        4: { name: '无法更换屏幕内容', id: '4' },
        5: { name: '屏幕移位', id: '5' },
        0: { name: '其他', id: '0' }
    },
};
let demand_cate = {
    "1": '安装',
    "2": '维修'
}

export function GetInitData(rootRequest: potato.RootRequest, data: { demand?: string, token?: string}, success?: (data: init.InitData) => void, fail?: (error: potato.PError) => void): Promise<init.InitData> {
    if (data.token) {
        rootRequest.setCookie(ApiDomain + '$token', data.token);
        rootRequest.assignCookie(ApiDomain + '$token', data.token);
    }
    return GetIpLocation(rootRequest, {}).then(function (ipInfo) {
        // 在cookie中存入当前ip所在省份
        rootRequest.setCookie(ApiDomain + '$ip_prov', ipInfo.province.province_name);
        return Promise.all([
            GetInit(rootRequest, { prov: ipInfo.province.province_name }),
            GetCurDemandInfo(rootRequest, data.demand),
        ]).then(function (list: any) {
            let city = ipInfo.city;
            let province = ipInfo.province;
            let selected_province = list[0].selected_province;
            let userInfo: user.UserInfo = list[0].user;
            let demand: Demand.DemandInfo[] = list[0].my_demand_list;//自己发布的项目列表
            let supList = list[0].partic_supplier_list;
            let supplier_cnt = list[0].supplier_cnt;
            let current_demand = list[1];
            var currPage = (rootRequest.args as any)["__content__"].split('/m')[1];// 当前页面或者模块
            if(currPage == '/'){currPage = '/Index'};
            let result: init.InitData = {
                user: userInfo,
                city: city,
                province: province,
                selected_city: city,
                selected_province: selected_province,
                demand: demand,
                current_demand: current_demand, //当前打开的项目详情项目
                demand_config: demand_config,
                project_list: [],
                currIsSelfDemand: false,
                supplier_cnt: supplier_cnt,
                unread_num: '0',
                supList:supList,
                currPage: currPage
            };
            success && success(result);
            return result;
        }).catch(fail || emptyFailHandler);
    }, function () {

    })
}
function mappingUser(json: any): user.UserInfo {
    json.user_type = parseInt(json.type);
    json.user_id = json.id;
    json.sex = json.gender || 1;
    json.nick = json.nickname || '';
    json.address = json.addr || '';
    json.balance = json.bal || '';
    json.position = json.pos || '';
    json.company_name = json.com || '';
    json.activated = json.active || '';
    json.phone = json.mobile || '';
    delete json.type;
    delete json.id;
    delete json.gender;
    delete json.pos;
    delete json.com;
    delete json.addr;
    delete json.bal;
    delete json.nickname;
    delete json.active;
    delete json.mobile;
    return json;
}
function mappingDemand(json: any) {
    // if(json.length == 0){
    //     return 
    // }
    let demand = Object.assign({}, json, json.demand);
    delete demand.demand;
    if (!demand.cate) { demand.cate = "1" };
    demand.cate_name = demand_cate[demand.cate];
    demand.type_name = demand_config.typeOptions[demand.type].name;
    demand.location_name = demand_config.locationOptions[demand.surr].name;
    demand.color_name = demand_config.colorOptions[demand.color].name;
    demand.malf_name = demand_config.faultOptions[demand.malf].name;
    demand.location = demand.surr;
    demand.image = demand.img;
    demand.note = demand.remark;
    demand.intm = parseInt(demand.created_at);
    demand.uptm = parseInt(demand.updated_at);
    demand.address = demand.addr;
    demand.supplier_list = demand.partic_supplier_list;
    demand.user.sex = demand.user.gender;
    demand.user.nick = demand.user.nickname;
    demand.supplier_list.forEach(function (item) {
        item.company_name = item.com;
        item.nick = item.nickname;
        item.unread = '';
    })
    demand.im_unread = {
        group: 0,
        friends: 0
    };
    demand.budgetFormat = budget(demand.budget);
    delete demand.surr;
    delete demand.img;
    delete demand.addr;
    delete demand.remark;
    delete demand.created_at;
    delete demand.updated_at;
    delete demand.partic_supplier_list;
    delete demand.gender;
    delete demand.user.mobile;
    delete demand.user.nickname;
    delete demand.user.gender;
    return demand;
}
export function GetIpLocation(rootRequest: potato.RootRequest, data: { ip?: string }, success?: (data: any) => void, fail?: (error: potato.PError) => void): Promise<any> {
    data.ip = data.ip || rootRequest.getIP();
    return potato.callApi({
        url: apiServer + '/Ip/getIpInfo', data: data, method: "POST", context: rootRequest, render: function (data) {
            let result = {
                city: {
                    city_id: '北京市',
                    city_name: '北京市'
                },
                province: {
                    province_id: '北京市',
                    province_name: '北京市'
                }
            };
            data = JSON.parse(data);
            if (data.status == 200) {
                // ip定位城市信息
                result = {
                    city: {
                        city_id: data.ip_info.city,
                        city_name: data.ip_info.city
                    },
                    province: {
                        province_id: data.ip_info.prov,
                        province_name: data.ip_info.prov
                    }
                }
            }
            return result
        }
    })
}
export function GetDemandListByTypes(rootRequest: potato.RootRequest, success?: (data: Demand.DemandInfo[][]) => void, fail?: (error: potato.PError) => void): Promise<Demand.DemandInfo[][]> {
    // 默认参数  
    var types = [2, 5, 4, 3, 1, 0];
    var time = new Date().getTime();
    var data: any = { type: types, sort: "created_at", sort_val: time, cnt: 8 }
    // 先查询是否有选择的城市的cookie
    let cookies: any = rootRequest.getCookie();
    let ipProv = cookies[ApiDomain + '$ip_prov'];
    data.prov = ipProv;
    console.log(JSON.stringify(data));
    return potato.callApi({
        url: apiServer + '/Demand/getDistrDemandList', data: data, method: "POST", context: rootRequest, render: function (data) {
            let result: Demand.DemandInfo[][], json: { status: number, info: string, list: { [key: string]: Demand.DemandInfo[] } };
            if (data instanceof potato.PError) {
                json = { status: 500 } as any;
            } else {
                json = JSON.parse(data);
            }
            if (json.status == 200) {
                result = types.map(function (type) {
                    let list = json.list[type];
                    if (list) {
                        return list.map(function (item) {
                            return mappingDemand(item);
                        })
                    } else {
                        return [];
                    }
                })
            } else {
                result = types.map(function (type) {
                    return [];
                });
            }

            return result;
        }
    }, success, fail);
}
export function GetInit(rootRequest: potato.RootRequest, data: { prov: string }, success?: (data: any) => void, fail?: (error: potato.PError) => void): Promise<any> {
    // 先查询是否有选择的城市的cookie
    let ipProv = data.prov;
    let cookies: any = rootRequest.getCookie();
    let selProv = cookies[ApiDomain + '$selected_prov'];
    // cookie中是否存在选择的城市
    let cookie_slected_prov: any = '';
    if (selProv) {
        data = {
            prov: selProv,
        };
        cookie_slected_prov = {
            province_id: selProv,
            province_name: selProv
        }
    }
    return potato.callApi({
        url: apiServer + '/Index/getInitData', data: data, method: "POST", context: rootRequest, render: function (data) {
            let result, json: any = {};
            if (data instanceof potato.PError) {
                json = { status: 500 } as any;
            } else {
                json = JSON.parse(data);
            }
            if (json.status == 200) {
                json.selected_province = {
                    province_id: ipProv,
                    province_name: ipProv
                }
                // 如果cookie中存在已选择的城市
                if (cookie_slected_prov) {
                    json.selected_province = cookie_slected_prov;
                }
                // 用户信息
                json.user = mappingUser(json.user);
                // 自己发布的项目信息
                if (json.my_demand_info == null) { json.my_demand_info = demand_empty } else { json.my_demand_info = mappingDemand(json.my_demand_info); }
                // 全国和当前省份的工程商数量
                json.supplier_cnt.prov = Math.round(json.supplier_cnt.prov / 50);
                json.supplier_cnt.total = Math.round(json.supplier_cnt.total / 50);
                // 参与的项目列表
                delete json.partic_demand_list;
                // 已开通的增值服务列表
                delete json.supplier_vas;
                //自己的项目列表
                
                if(json.my_demand_list){
                    let num = json.my_demand_list.length;
                    for (let i = 0; i < num; i++) {
                        json.my_demand_list[i] = mappingDemand(json.my_demand_list[i]);
                    }
                }else{
                    json.my_demand_list=[];
                }
                // 组合参与的工程商列表
                var supplier_list: any = [];
                var demandList = json.my_demand_list;
                if (demandList.length == 1) {
                    supplier_list = demandList[0].supplier_list;
                } else {
                    var list = {};
                    demandList.forEach(function (demand) {
                        demand.supplier_list.forEach(function (partic_supplier_list) {
                            list[partic_supplier_list.id] = partic_supplier_list;
                        })
                    });
                    for (var key in list) {
                        supplier_list.push(list[key]);
                    }
                }
                json.partic_supplier_list = supplier_list;
                result = json;
            }else{
                json.my_demand_list=[];
                result = json;
            }
            
            return result;
        }
    }, success, fail);
}
export function GetCurDemandInfo(rootRequest: potato.RootRequest, demand_id?: string, success?: (data: Demand.DemandInfo) => void, fail?: (error: potato.PError) => void): Promise<Demand.DemandInfo> {
    if (!demand_id) {
        success && success(demand_empty);
        return Promise.resolve(demand_empty);
    }
    return potato.callApi({
        url: apiServer + '/Demand/getInfo', data: { id: demand_id }, method: "POST", context: rootRequest, render: function (data) {
            let result: Demand.DemandInfo, json: { status: number, info: string };
            if (data instanceof potato.PError) {
                json = { status: 500 } as any;
            } else {
                json = JSON.parse(data);
            }
            if (json.status == 200 && (json as any).info) {
                result = mappingDemand((json as any).info);
            } else {
                result = demand_empty;
            }
            return result;
        }
    }, success, fail);
}
