// views/m/project/InqTypeSelect/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as tdom from "@po-to/tomato-jquery";
import * as Vue from "vue";
import code = require('views/m/user/Code');

class VPresenter extends project.VPresenter {
    private errorTip: JQuery;
    private errorData: JQuery;
    private typeId: any;
    private selected_city;
    private selected_prov;
    private span;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this.errorTip = this.find(".errorTip");
        this.typeId = 2;
        var selected_city = this.selected_city = model.globalData.selected_city;
        this.selected_prov = model.globalData.selected_province;
        var that = this;
        var inquiryTypeSelect = this.find(".citySelect").get(0);
        var pro: any = new Vue({
            el: inquiryTypeSelect,
            data: {
                selected_city: selected_city,
            },
        });
        // 填充名字
        var nick = model.globalData.user.nick || '';
        that.find('.nickName').val(nick);

        $('input').bind("input propertychange", function () {
            that.errorTip.addClass("hide");
        });
        //选择类型
        $(".choice-type").each(function (index, element) {
            var that_son = $(this);
            that_son.on("click", function () {
                $(".select").removeClass("isSelected");
                that_son.find('.select').addClass("isSelected");
                that.typeId = that_son.find('.select').attr('type');
                // that.updateInit(choiceType.find('.select').attr('type'));
            });
        });
        //屏幕类型下一步
        $(".inquiry-type-next").on("click", function () {
            $("#TypeSelect").addClass("hide");
            $("#ColorSelect").removeClass("hide");
            $("#LastSelect").addClass("hide");
            that.updateInit(that.typeId);
        });
        //颜色下一步
        $(".inquiry-color-next").on("click", function () {
            if (that.checkSpan('span', "间距") && that.checkSize(that.find('.size').val(), 'size', "面积")) {
                $("#TypeSelect").addClass("hide");
                $("#ColorSelect").addClass("hide");
                $("#LastSelect").removeClass("hide");
            }
        });
        //颜色上一步
        $(".inquiry-color-prv").on("click", function () {
            $("#TypeSelect").removeClass("hide");
            $("#ColorSelect").addClass("hide");
            $("#LastSelect").addClass("hide");
        });
        //last上一步
        $(".inquiry-last-prv").on("click", function () {
            $("#TypeSelect").addClass("hide");
            $("#ColorSelect").removeClass("hide");
            $("#LastSelect").addClass("hide");
        });
        this._watchEvent();
        this._els = this._getElements();
    }
    /**间距校验 */
    private checkSpan(dom, tip) {
        //var reg = /^\d+(\.\d+)?$/
        if (this.find(".custom-options-button").hasClass("hide")) {
            this.span = this.find('.span').val()
        } else {
            this.span = this.find('.spanChoosed').text();
        }
        if (this.span == "") {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-span').text(tip + '不能为空！')
            return false;
            //} else if (!reg.test(info)) {
            //    $('.' + dom + 'Tip').removeHide();
            //    $('.' + dom + 'Tip').find('.error-span').text('请输入正确的' + tip);
            //    return false;
        } else {
            return true;
        }
    }
    /**面积校验 */
    private checkSize(info, dom, tip) {
        var reg = /^\d+(\.\d+)?$/
        if (info == "") {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-size').text(tip + '不能为空！')
            return false;
        } else if (!reg.test(info)) {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-size').text('请输入正确的' + tip);
            return false;
        } else {
            return true;
        }
    }
    /**昵称校验 */
    private checkNick(info, dom, tip) {
        var reg = /^[\u4e00-\u9faf]+$/
        if (info == "") {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-nick').text(tip + '不能为空！')
            return false;
        } else if (!reg.test(info)) {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-nick').text('请输入正确的' + tip);
            return false;
        } else {
            return true;
        }
    }
    /**本来想写一个来着，但是取值不好取，就写了三个 */
    private _evt_select1 = function (data, target, hit) {
        this.locSelect(target);
        return false;
    }
    private _evt_select2 = function (data, target, hit) {
        this.corSelect(target);
        return false;
    }
    private _evt_select3 = function (data, target, hit) {
        this.spanSelect(target);
        return false;
    }

    /**实现选择屏幕联动改变颜色、位置的初始值 */
    private updateInit(scrType) {
        //var list = $("#InquiryColorSelect").find('.choose');
        var list = this.find('.choose');

        for (var i = 0; i < list.length; i++) {
            if (scrType == 0) {
                if ($(list[i]).attr('loc') == "1") {
                    this.locSelect(list[i]);
                }
                if ($(list[i]).attr('color') == "3") {
                    this.corSelect(list[i]);
                }
                this.difSpan('P12', 'P10', 'P8')
                if ($(list[i]).text() == "P10") {
                    this.spanSelect(list[i]);
                }
            } else if (scrType == 1) {
                if ($(list[i]).attr('loc') == "3") {
                    this.locSelect(list[i]);
                }
                if ($(list[i]).attr('color') == "1") {
                    this.corSelect(list[i]);
                }
                this.difSpan('P12', 'P10', 'P8')
                if ($(list[i]).text() == "P10") {
                    this.spanSelect(list[i]);
                }
            } else if (scrType == 2) {
                if ($(list[i]).attr('loc') == "1") {
                    this.locSelect(list[i]);
                }
                if ($(list[i]).attr('color') == "3") {
                    this.corSelect(list[i]);
                }
                this.difSpan('P6', 'P8', 'P10')
                if ($(list[i]).text() == "P10") {
                    this.spanSelect(list[i]);
                }
            } else if (scrType == 3) {
                if ($(list[i]).attr('loc') == "1") {
                    this.locSelect(list[i]);
                }
                if ($(list[i]).attr('color') == "1") {
                    this.corSelect(list[i]);
                }
                this.difSpan('P6', 'P8', 'P10')
                if ($(list[i]).text() == "P10") {
                    this.spanSelect(list[i]);
                }
            } else if (scrType == 4) {
                if ($(list[i]).attr('loc') == "2") {
                    this.locSelect(list[i]);
                }
                if ($(list[i]).attr('color') == "3") {
                    this.corSelect(list[i]);
                }
                this.difSpan('P3', 'P4', 'P5')
                if ($(list[i]).text() == "P4") {
                    this.spanSelect(list[i]);
                }
            } else {
                if ($(list[i]).attr('loc') == "2") {
                    this.locSelect(list[i]);
                }
                if ($(list[i]).attr('color') == "3") {
                    this.corSelect(list[i]);
                }
                this.difSpan('P1.25', 'P2.0', 'P2.5')
                if ($(list[i]).text() == "P2.0") {
                    this.spanSelect(list[i]);
                }
            }
        }
    }
    // 
    /**选择不同类型屏幕对应默认不同的间距*/
    private difSpan(p1, p2, p3) {
        $($('.span3 .choose')[0]).text(p1)
        $($('.span3 .choose')[1]).text(p2)
        $($('.span3 .choose')[2]).text(p3)
    }
    /**选择一种类型之后清除其他选项的底色，三个方法 */
    private locSelect(dom) {
        $(dom).parent().parent().find('.choose').removeClass('locChoosed');
        $(dom).addClass('locChoosed')
    }
    private corSelect(dom) {
        $(dom).parent().parent().find('.choose').removeClass('corChoosed');
        $(dom).addClass('corChoosed')
    }
    private spanSelect(dom) {
        $(dom).parent().parent().find('.choose').removeClass('spanChoosed');
        $(dom).addClass('spanChoosed')
    }
    /**inquirySubmit */
    _evt_inquirySubmit(data, target, hit) {
        // $(target).attr('evt', '');
        var that = this;
        var mobile = model.globalData.user.phone;
        var cate = 1;//安装
        var sem = localStorage.getItem("repair_statistics");
        var type = that.find('.isSelected').attr('type');
        var size = that.find('.size').val();
        var location = that.find('.locChoosed').attr('loc');
        var color = that.find('.corChoosed').attr('color');
        var nick = that.find('.nickName').val();
        var city = that.find('.citySelect').text();
        var prov = that.selected_prov.province_name;
        var addr = that.find(".addr").val();
        var span = that.span;
        var prePostInfo = {
            mobile: mobile,
            cate: cate,
            prov: prov,
            city: city,
            sem: sem
        }
        if (that.checkNick(nick, 'nick', "称呼")) {

            api.PrePost(prePostInfo, function (data: any) {
                var demand_id = data.id;
                var InquiryInfo = {
                    id: demand_id,
                    nickname: nick,
                    type: type,
                    surr: location,
                    size: size,
                    color: color,
                    span: span,
                    prov: prov,
                    city: city,
                    addr: addr,
                }
                api.ActivateDemand(InquiryInfo, function (data: any) {
                    //发布成功，，直接打开聊天室（“已登录，无在建项目”，询价）
                    window.location.reload();
                    // window.location.href = '/ChatRoom/?demand=' + data.demand_info.demand.id;
                }, function (error: tomato.PError) {
                    if (error.name == "401") {
                        //未登录
                        funs.tip(error.note, "", "fail");
                        setTimeout(function () {
                            window.location.reload();
                        }, 2000);
                    } else if (error.name == "41004002") {

                        //项目不存在
                        funs.tip(error.note, "", "fail");
                        setTimeout(function () {
                            window.location.reload();
                        }, 2000);
                    } else {
                        funs.tip(error.note, "", "fail");
                        setTimeout(function () {
                            window.location.reload();
                        }, 2000);
                    }
                });

            }, function (error: tomato.PError) {
                if (error.name == '401') {
                    //未登录，需弹出验证码弹窗进行登录
                    tomato.getVPresenter<code>(funs.moduleToUrl('m/user/Code'), (code) => {
                        code.setCallBack(mobile, function () { that._evt_inquirySubmit(data, target, hit); });
                        tdom.open(code, tdom.DialogTarget.Blank, {
                            size: { width: '90%', height: tomato.DialogSize.Content },
                            masked: true,
                        });
                    });
                } else {
                    funs.tip(error.name, error.note, "fail");
                }
            });
        } else {
            // $(target).attr('evt', 'inquirySubmit');
        }
        return false;
    }
    /**城市选择 */
    _evt_citySelect() {
        // tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/CitySelector'), (CitySelector) => {
        //     (CitySelector as any).setProv("Repair");
        //     funs.loadPage(CitySelector);
        // })
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/CitySelector'), (CitySelector) => {
            //登录窗
            (CitySelector as any).setProv("select");
            tdom.open(CitySelector, tdom.DialogTarget.Blank, {
                size: { width: '100%', height: "100%" },
                effect: tdom.DialogEffect.scale,
                bodyEffect: tdom.TurnEffect.cover,
                masked: true
            });
        });
    }
    /**报装间距自定义 */
    _evt_customOptionsInput() {
        this.find(".custom-options-button").addClass("hide");
        this.find(".custom-options-input").removeClass("hide");
    }
    /**报装间距返回 */
    _evt_customOptionsChoose() {
        this.find(".custom-options-input").addClass("hide");
        this.find(".custom-options-button").removeClass("hide");
    }
}
export = VPresenter;
