// views/m/joinSupplier/SupplierInfo/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";
import * as tdom from "@po-to/tomato-jquery";
import code = require('views/m/user/Code');

class VPresenter extends project.VPresenter {
    /**模块 */
    public SupplierInfo_vue: any;
    /**user */
    private user;
    private joinData_vue: {
        user: {},
        checkPhoneData: string,
        phoneError: Boolean,
        nickError: Boolean,
        clicked: Boolean,
        phoneEdit: Boolean
    };
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this.joinData_vue = {
            user: {},
            checkPhoneData: '',
            phoneError: false,
            nickError: false,
            clicked: true,
            phoneEdit: false
        };
        var vueDom = this.find("#SupplierInfo_vue").get(0);
        var user = this.user = model.globalData.user;
        var city = model.globalData.city;
        let that = this;
        this.getGlobalData();
        let SupplierInfo_vue: any = this.SupplierInfo_vue = new Vue({
            el: vueDom,
            data: that.joinData_vue,
            watch: {
                'user.phone': "checkData",
                'user.nick': "checkData",
                'user.address': "checkData",
                'user.company_name': "checkData",
                'user.position': "checkData",

            },
            methods: {
                //限制输入
                checkData: function () {
                    SupplierInfo_vue.user.phone = SupplierInfo_vue.user.phone.replace(/[^\d]/g, '');
                    SupplierInfo_vue.user.nick = SupplierInfo_vue.user.nick.replace(/[^(\u4e00-\u9fa5)]/, '');
                    SupplierInfo_vue.user.address = SupplierInfo_vue.user.address.replace(/[^(\u4e00-\u9fa5)]/, '');
                    SupplierInfo_vue.user.company_name = SupplierInfo_vue.user.company_name.replace(/[^(\u4e00-\u9fa5)]/, '');
                    SupplierInfo_vue.user.position = SupplierInfo_vue.user.position.replace(/[^(\u4e00-\u9fa5)]/, '');
                    SupplierInfo_vue.checkInfo();
                    SupplierInfo_vue.$watch(user.phone, 'checkData');
                },
                //检验必填项
                checkInfo: function () {
                    if ("" == SupplierInfo_vue.user.nick) {
                        SupplierInfo_vue.nickError = true;
                        SupplierInfo_vue.clicked = true;
                        return false;
                    } else {
                        SupplierInfo_vue.nickError = false;
                        if (SupplierInfo_vue.checkPhone(SupplierInfo_vue.user.phone) && SupplierInfo_vue.user.phone && SupplierInfo_vue.user.nick && model.globalData.city.city_name && model.globalData.province.province_name) {
                            SupplierInfo_vue.clicked = false;
                            return true;
                        } else {
                            SupplierInfo_vue.clicked = true;
                            return false;
                        }
                    }

                },
                //正则校验手机号
                checkPhone: function (phone: string): boolean {
                    if (phone) {
                        if (!(/^1(3|4|5|7|8)\d{9}$/.test(phone))) {
                            SupplierInfo_vue.checkPhoneData = '请输入正确的手机号';
                            SupplierInfo_vue.phoneError = true;
                            return false;
                        } else {
                            SupplierInfo_vue.phoneError = false;
                            return true;
                        }
                    } else {
                        SupplierInfo_vue.checkPhoneData = '手机号不能为空';
                        return false;
                    }

                },
                submitEdit: function () {
                    console.log(SupplierInfo_vue.user.nick + "::" + SupplierInfo_vue.user.phone + "::" + SupplierInfo_vue.user.address + "::" + SupplierInfo_vue.user.company_name + "::" + SupplierInfo_vue.user.position);
                    //点击提交
                    let nickname = SupplierInfo_vue.user.nick;
                    let mobile = SupplierInfo_vue.user.phone
                    let gender = SupplierInfo_vue.user.sex;
                    let prov = model.globalData.province.province_name;
                    let city = model.globalData.city.city_name;
                    let addr = SupplierInfo_vue.user.address ? SupplierInfo_vue.user.address : undefined;
                    let com = SupplierInfo_vue.user.company_name ? SupplierInfo_vue.user.company_name : undefined;
                    let pos = SupplierInfo_vue.user.position ? SupplierInfo_vue.user.position : undefined;
                    let userInfo = {
                        nickname: nickname,
                        mobile: mobile,
                        gender: gender,
                        prov: prov,
                        city: city,
                        addr: addr,
                        com: com,
                        pos: pos
                    }
                    $.each(userInfo, function (key, value) {
                        if (value==undefined) {
                            delete userInfo[key];
                        }
                    });
                    //已登录,调用加盟接口
                    api.SupplierJoin(userInfo, function () {
                        //成功，刷新
                        that.setGlobalData();
                        window.location.href = '/index.html';
                    }, function (error: tomato.PError) {
                        if (error.name == '401') {
                            //未登录，去执行登录
                            //弹出验证码
                            tomato.getVPresenter<code>(funs.moduleToUrl('m/user/Code'), (code) => {
                                code.setCallBack(userInfo.mobile, function () { SupplierInfo_vue.submitEdit() });
                                tdom.open(code, tdom.DialogTarget.Blank, {
                                    size: { width: '90%', height: tomato.DialogSize.Content },
                                    masked: true,
                                });
                            });
                        } else if (error.name == '41009003') {
                            //已是工程商
                            funs.tip(error.note, "", "fail");
                            setTimeout(function () {
                                window.location.href = '/index.html';
                            }, 3020);
                        } else {
                            funs.tip(error.name, error.note, "fail");
                        }
                    });
                }
            }
        });
        //this._watchEvent();
        //this._els = this._getElements();

    }
    public init() {
        if (1 == this.SupplierInfo_vue.user.user_type || 100 == this.SupplierInfo_vue.user.user_type) {
            //已登录，手机号不可编辑
            this.SupplierInfo_vue.phoneEdit = true;
        } else {
            //未登录，手机号可编辑
            this.SupplierInfo_vue.phoneEdit = false;
        }
    }
    _afterInstallTo() {
        //加载模块时
        this.SupplierInfo_vue.checkData();
        this.SupplierInfo_vue.nickError = false;
        this.SupplierInfo_vue.phoneError = false;
    }
    _afterUninstallTo() {
        //退出时
        this.getGlobalData();
    }
    setGlobalData() {
        //this.vueData => globalData
        $.extend(true, model.globalData.user, this.joinData_vue.user);

    }
    getGlobalData() {
        //globalData => this.vueData
        this.joinData_vue.user = funs.deepCopy(model.globalData.user);
    }
}
export = VPresenter;
