// views/m/DemandDetail/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as api from "views/global/common/API";
import * as model from "views/global/common/Model";
import * as Vue from 'vue';
import edit = require('views/m/project/EditCurrentItem');

class VPresenter extends project.VPresenter {
    private proInfo: any;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._watchEvent();
        this._els = this._getElements();
        var that = this;
        //详情页渲染
        var v_details = this.find('.details')[0];
        var data = model.globalData;
        this.proInfo = new Vue({
            el: v_details,
            data: data,
            methods: {
                openEdit: function () {
                    var moduleUrl = 'm/RepairEditCurrent';
                    if (model.globalData.current_demand.cate == "1") {
                        //安装
                        moduleUrl = 'm/project/EditCurrentItem';
                    }
                    tomato.getVPresenter<edit>(funs.moduleToUrl(moduleUrl), (edit) => {
                        var type = model.globalData.current_demand.type;
                        edit.setInit(type);
                        funs.loadPage(edit);
                    })
                }
            }
        });
    }
    /*==============录音音频处理-start=============*/
    //录音播放暂停
    _evt_playPause(data, target) {
        var that = this;
        //如果录音加载失败
        var audioTime = this.find('.audio-time');
        var audio: any = document.getElementById("audioTag");
        if (audio.paused) {
            audio.play();
            //改变icon
            $(target).find('span').removeClass('icon-play-one').addClass('icon-pause-one')
        } else {
            audio.pause();
            $(target).find('span').removeClass('icon-pause-one').addClass('icon-play-one')
        }

        //监听音频播放时间并更新进度条
        audio.addEventListener('timeupdate', that.updateProgress, false);
        //监听播放完成事件
        audio.addEventListener('ended', that.audioEnded, false);
    };

    //更新进度条
    updateProgress() {
        var audio = document.getElementsByTagName('audio')[0];
        this.find('.audio-time').html(this.transTime(audio.duration - audio.currentTime));
    }
    //播放完成
    audioEnded() {
        var audio = document.getElementsByTagName('audio')[0];
        audio.currentTime = 0;
        audio.pause();
        this.find('.play-pause>span').removeClass('icon-pause-one').addClass('icon-play-one');
    }
    //转换音频时长格式
    transTime(time) {
        var duration = parseInt(time);
        var minute: any = Math.floor(duration / 60);
        var sec = duration % 60 + '';
        var isM0 = ':';
        if (minute == 0) {
            minute = '00';
        } else if (minute < 10) {
            minute = '0' + minute;
        }
        if (sec.length == 1) {
            sec = '0' + sec;
        }
        return minute + isM0 + sec
    }
    /*==============录音音频处理-end=============*/

    //点击查看大图
    _evt_scanBig = function (data, obj) {
        var src = $(obj).attr('src');
        var width = $(obj).width();
        var height = $(obj).height();
        var rate = width / height;//宽高比
        var hh = $(window).width();//大图的宽 == 屏幕宽度
        var bgh = hh / rate / 2;  // 计算出大图的高度 de 一半

        //dom插入
        var bigImg = '<div class="bigImgBg" evt="closeBigImg"><img class="bigImg" src="' + src + '"/></div>';
        $(this.view).append(bigImg);

        //让图片垂直居中
        $('.bigImg').css('margin-top', -bgh + 'px');
    };
    _evt_closeBigImg = function (data, obj) {
        $(obj).remove();
    };
}
export = VPresenter;
