// views/m/project/EditCurrentItem/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as Vue from "vue";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as funs from "views/global/common/Funs";
import ChatRoom = require('views/m/ChatRoom');
import ImageUploader = require('views/m/project/ImageUploader');

class VPresenter extends project.VPresenter {
    /**模块 */
    private EditCurrentItem_vue: any;
    /**循环选择的数据源 */
    private demand_config;
    private img;

    private vueData: {
        typeClick: boolean,
        locationClick: boolean,
        colorClick: boolean,
        spanClick: boolean,
        OtherClick: boolean,
        otherSpan: string
        clicked: boolean,
        user: any,
        demand: any,
        Url: {
            typeUrl: any[],
            locationUrl: any[],
            colorUrl: any[],
            spanUrl: any[]
        }

    };
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        var vueDom = this.find("#EditCurrentItem_vue").get(0);
        this.vueData = {
            typeClick: false,
            locationClick: false,
            colorClick: false,
            spanClick: false,
            OtherClick: false,
            otherSpan: '',
            clicked: true,
            user: {},
            demand: {},
            Url: {
                typeUrl: [],
                locationUrl: [],
                colorUrl: [],
                spanUrl: [],
            }
        };
        var that = this;
        let EditCurrentItem_vue: any = this.EditCurrentItem_vue = new Vue({
            el: vueDom,
            data: that.vueData,
            watch: {
                'demand.address': "checkData",
                'demand.size': "checkData",
                'demand.budget': "checkData"
            },
            methods: {
                checkData: function () {
                    EditCurrentItem_vue.demand.address = EditCurrentItem_vue.demand.address.replace(/[^(\u4e00-\u9fa5)]/g, '');
                    EditCurrentItem_vue.demand.size = (EditCurrentItem_vue.demand.size).toString().replace(/[^\d\.]/g, '');
                    EditCurrentItem_vue.demand.budget = (EditCurrentItem_vue.demand.budget).toString().replace(/[^\d\.]/g, '');
                    EditCurrentItem_vue.checkInfo();
                },
                checkInfo: function () {
                    //检验必填项
                    if (EditCurrentItem_vue.demand.address && EditCurrentItem_vue.demand.size) {
                        EditCurrentItem_vue.clicked = false;
                    } else {
                        EditCurrentItem_vue.clicked = true;
                    }
                },
                //展开下拉菜单
                showType: function () {
                    EditCurrentItem_vue.typeClick = !EditCurrentItem_vue.typeClick;
                    EditCurrentItem_vue.locationClick = false;
                    EditCurrentItem_vue.colorClick = false;
                    EditCurrentItem_vue.spanClick = false;
                },
                showLocation: function () {
                    EditCurrentItem_vue.locationClick = !EditCurrentItem_vue.locationClick;
                    EditCurrentItem_vue.typeClick = false;
                    EditCurrentItem_vue.colorClick = false;
                    EditCurrentItem_vue.spanClick = false;
                },
                showColor: function () {
                    EditCurrentItem_vue.colorClick = !EditCurrentItem_vue.colorClick;
                    EditCurrentItem_vue.typeClick = false;
                    EditCurrentItem_vue.locationClick = false;
                    EditCurrentItem_vue.spanClick = false;
                },
                showSpan: function () {
                    if (EditCurrentItem_vue.OtherClick) {
                        EditCurrentItem_vue.showCloseAll();
                    } else {
                        EditCurrentItem_vue.spanClick = !EditCurrentItem_vue.spanClick;
                        EditCurrentItem_vue.typeClick = false;
                        EditCurrentItem_vue.locationClick = false;
                        EditCurrentItem_vue.colorClick = false;
                    }

                },
                showCloseAll: function () {
                    EditCurrentItem_vue.spanClick = false;
                    EditCurrentItem_vue.typeClick = false;
                    EditCurrentItem_vue.locationClick = false;
                    EditCurrentItem_vue.colorClick = false;
                    EditCurrentItem_vue.OtherClick = false;
                },
                //参数选择
                getType: function (event) {
                    let id = event.target.getAttribute('value');
                    let name = event.target.innerHTML;
                    that.setTpye(id);

                    EditCurrentItem_vue.showType();

                },
                getLocation: function (event) {
                    EditCurrentItem_vue.demand.location = event.target.getAttribute('value');
                    EditCurrentItem_vue.demand.location_name = event.target.innerHTML;
                    EditCurrentItem_vue.showLocation();
                },
                getColor: function (event) {
                    EditCurrentItem_vue.demand.color = event.target.getAttribute('value');
                    EditCurrentItem_vue.demand.color_name = event.target.innerHTML;
                    EditCurrentItem_vue.showColor();
                },
                getSpan: function (event) {
                    let id = event.target.getAttribute('value');
                    let name = event.target.innerHTML;
                    // EditCurrentItem_vue.demand.span = name;
                    EditCurrentItem_vue.demand.span_id = id;
                    if (4 == id) {
                        EditCurrentItem_vue.showCloseAll();
                        EditCurrentItem_vue.OtherClick = true;
                        EditCurrentItem_vue.otherSpan = '';
                    } else {
                        EditCurrentItem_vue.demand.span = name;
                        EditCurrentItem_vue.showSpan();
                    }
                },
                getSure: function () {
                    EditCurrentItem_vue.demand.span = EditCurrentItem_vue.otherSpan;
                    EditCurrentItem_vue.showCloseAll();
                },
                getClean: function () {
                    // EditCurrentItem_vue.otherSpan = '';
                    EditCurrentItem_vue.showCloseAll();
                    setTimeout(function () {
                        EditCurrentItem_vue.showSpan();
                    }, 0);

                }, /**提交 */
                submitEdit: function () {
                    var demand_id = parseInt(EditCurrentItem_vue.demand.id);
                    var type = parseInt(EditCurrentItem_vue.demand.type);
                    var size = parseFloat(EditCurrentItem_vue.demand.size);
                    var color = parseInt(EditCurrentItem_vue.demand.color);
                    var location = parseInt(EditCurrentItem_vue.demand.location);
                    var span = EditCurrentItem_vue.demand.span;
                    var budget = EditCurrentItem_vue.demand.budget ? parseFloat(EditCurrentItem_vue.demand.budget) : 0;
                    var address = EditCurrentItem_vue.demand.address;
                    var image = EditCurrentItem_vue.demand.image;
                    var note = EditCurrentItem_vue.demand.note ? EditCurrentItem_vue.demand.note : undefined;
                    let editInfo = {
                        id: demand_id,
                        type: type,
                        size: size,
                        color: color,
                        surr: location,
                        budget: budget,
                        span: span,
                        addr: address,
                        remark: note
                    }
                    if ("data:image/jpeg;base64," == model.globalData.current_demand.image.substring(0, 23)) {
                        //提交图片
                        tomato.getVPresenter<ImageUploader>(funs.moduleToUrl('m/project/ImageUploader'), (ImageUploader) => {
                            ImageUploader.fileUpload();
                        });
                    }

                    api.EditDemand(editInfo, function (data: any) {
                        // window.location.reload();
                        EditCurrentItem_vue.demand.budgetFormat.budget = api.budget(budget).budget;
                        EditCurrentItem_vue.demand.budgetFormat.unit = api.budget(budget).unit;
                        that.setGlobalData();
                        // model.globalData.current_demand = data.demand_info.demand;
                        funs.viewHistory.go(-1);
                    }, function (error: tomato.PError) {
                        console.log(error);
                        funs.tip(error.note, "", "fail");
                    });
                },
            }
        });
        this._watchEvent();
    }
    /**编辑项目时，将demand的type传过来，以便做样式上的初始化 */
    public setInit(typeId) {
        this.getGlobalData();
        this.setTpye(typeId);
    }
    public setTpye(typeId) {
        let def = this.demand_config.typeOptions[typeId].def;
        let name = this.demand_config.typeOptions[typeId].name;

        this.EditCurrentItem_vue.demand.type = typeId;
        this.EditCurrentItem_vue.demand.type_name = name;

        this.EditCurrentItem_vue.Url.locationUrl = this.demand_config.locationOptions = def.locationOptions;
        this.EditCurrentItem_vue.Url.colorUrl = this.demand_config.colorOptions = def.colorOptions;
        this.EditCurrentItem_vue.Url.spanUrl = this.demand_config.spanOptions = def.spanOptions;


        this.EditCurrentItem_vue.demand.location = def.location;
        this.EditCurrentItem_vue.demand.location_name = this.EditCurrentItem_vue.Url.locationUrl[def.location].name;
        this.EditCurrentItem_vue.demand.color = def.color;
        this.EditCurrentItem_vue.demand.color_name = this.EditCurrentItem_vue.Url.colorUrl[def.color].name;
        this.EditCurrentItem_vue.demand.span = this.EditCurrentItem_vue.Url.spanUrl[def.span].name;
        this.EditCurrentItem_vue.demand.span_id = def.span;

        //检测必填项
        this.EditCurrentItem_vue.checkInfo();

    }
    setGlobalData() {
        //this.vueData => globalData
        var imgUrl = model.globalData.current_demand.image;
        $.extend(true, model.globalData.user, this.vueData.user);
        $.extend(true, model.globalData.current_demand, this.vueData.demand);
        $.extend(true, model.globalData.demand_config, this.demand_config);
        model.globalData.current_demand.image = imgUrl;

    }
    getGlobalData() {
        //globalData => this.vueData
        this.img = $(".imageUrl").attr("src");
        this.vueData.user = funs.deepCopy(model.globalData.user);
        this.vueData.demand = funs.deepCopy(model.globalData.current_demand);
        var demand_config = this.demand_config = funs.deepCopy(model.globalData.demand_config);
        this.vueData.Url = {
            typeUrl: [demand_config.typeOptions[2], demand_config.typeOptions[5], demand_config.typeOptions[3], demand_config.typeOptions[1], demand_config.typeOptions[4], demand_config.typeOptions[0],],
            locationUrl: [demand_config.locationOptions[1], demand_config.locationOptions[3], demand_config.locationOptions[2]],
            colorUrl: [demand_config.colorOptions[3], demand_config.colorOptions[2], demand_config.colorOptions[1]],
            spanUrl: [demand_config.spanOptions[1], demand_config.spanOptions[2], demand_config.spanOptions[3], demand_config.spanOptions[4]],
        };
        if (model.globalData.current_demand.budget == 0) {
            this.vueData.demand.budget = '';
        }
    }
}
export = VPresenter;
