// views/m/common/Header/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as model from "views/global/common/Model";
import * as funs from "views/global/common/Funs";
import * as api from "views/global/common/API";


class VPresenter extends project.VPresenter {
    private _menuDialog: tomato.Dialog | undefined;
    private _icon: JQuery;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._icon = this.find(".icon-circle-down");
        this._watchEvent();
        this._els = this._getElements();

        // 维修页修改文案
        if(model.globalData.currPage == "repair"){
            this.find('.panel-joined a').attr('href','/index.html').text('我要安装');
        }
    }
    _evt_showMenu(data: any, target: HTMLElement, hit: HTMLElement) {
        if (this._menuDialog) {
            this._menuDialog.close();
            this._icon.removeClass('rate');
        } else {
            this._icon.addClass('rate');
            tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/Menu'), (Menu) => {
                this._menuDialog = tdom.open(Menu, tdom.DialogTarget.Blank, {
                    className: "Dialog-Menu",
                    effect: tdom.DialogEffect.slideDown,
                    masked: true,
                    size: { width: tomato.DialogSize.Full, height: tomato.DialogSize.Content },
                    position: { x: 0, y: 0 },
                    offset: { x: 0, y: $(this.view[0]) }
                });
                this._menuDialog['_afterClose'] = () => {
                    this._menuDialog = undefined;
                    this._icon.removeClass('rate');
                }
            })
        }
    }
}
export = VPresenter;
