// views/m/project/InMedia/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as swiper from 'swiper';

class VPresenter extends project.VPresenter{
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?:string){
        super(view,parent,vpid);
        var a = swiper;
        this.callSwiper();
        //this._watchEvent();
        //this._els = this._getElements();
    }
     /**swiper */
    private callSwiper() {
        var mySwiper = new Swiper('.swiper-container', {
            direction: 'horizontal',
            // 如果需要分页器
            paginationClickable: true,
            pagination: '.swiper-pagination',

            // 如果需要前进后退按钮
            nextButton: '.swiper-button-next',
            prevButton: '.swiper-button-prev',
            noSwiping: true,
            //autoplay : 5000,
            //设置为true时，鼠标覆盖Swiper时指针会变成手掌形状，拖动时指针会变成抓手形状
            //grabCursor : true,
            slideToClickedSlide: true,//设置为true则点击slide会过渡到这个slide。
        })
    }
}
export = VPresenter;
