// nodejs/global/model/demand.ts

export interface supplier {
    id: string,
    nick: string,
    company_name: string,
    avatar: string,
    type: number,
    online: number,
    pptm: number,
    demand_customer_supplier_im_group_outer_id: string
}
export interface demandConfigItem {
    name: string,
    id: string,
}
export interface demandConfig {
    typeOptions: {
        0: any,
        1: any,
        2: any,
        3: any,
        4: any,
        5: any,
    },
    locationOptions: {
        1: demandConfigItem,
        2: demandConfigItem,
        3: demandConfigItem,
    },
    colorOptions: {
        1: demandConfigItem,
        2: demandConfigItem,
        3: demandConfigItem,
    },
    spanOptions: {
        1: demandConfigItem,
        2: demandConfigItem,
        3: demandConfigItem,
        4: demandConfigItem,
    }
    faultOptions: {
        1: demandConfigItem,
        2: demandConfigItem,
        3: demandConfigItem,
        4: demandConfigItem,
        5: demandConfigItem,
        0: demandConfigItem,
    },
}
export interface user {
    id: string,
    nick: string,
    sex: number
}
export interface DemandInfo {
    user: user,
    cate:string,
    cate_name:string,
    type_name: string,
    type: number,
    location_name: string,
    location: number,
    color_name: string,
    color: number,
    malf:string,
    malf_name:string,
    span: string,
    span_id: number,
    address: string,
    size: string,
    budget: number,
    budgetFormat:{budget:string,unit:string},
    id: string, //项目ID；
    city_id: string,
    city_name: string,
    prov:string,
    image: string,	//需求场地图片
    status: string, //10进行中，60: 交易结束
    intm: number,
    uptm: number,
    note: string,
    audio: string, //客服和客户之间的录音链接
    bid_fee: number, //抢单服务费
    supplier_list: supplier[],
    im_unread?:{
        group: number,
        friends: number
    }
}
export interface DemandList{
    [key:string]:DemandInfo
}
export interface DemandListByType{
    title:string,
    list: {double:DemandInfo[]}[]
}
