// views/m/RepairTypeSelect/index.ts
import * as tomato from "@po-to/tomato";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as tdom from "@po-to/tomato-jquery";
import * as Vue from "vue";
import code = require('views/m/user/Code');

class VPresenter extends project.VPresenter {
    private typeId;
    private fault;
    private errorTip;
    private nick;
    private repairId;
    private city;
    private addr;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this.typeId = 2;
        this.fault = 1;
        this.errorTip = this.find(".errorTip");
        this.nick = this.find(".nickName");
        this.city = this.find('.citySelect');
        this.addr = this.find(".addr");
        this.repairId;
        var selected_city = model.globalData.selected_city;
        var that = this;
        var repairTypeSelect = this.find(".citySelect").get(0);
        var pro: any = new Vue({
            el: repairTypeSelect,
            data: {
                selected_city: selected_city,
            },
        });
        $('input').bind("input propertychange", function () {
            that.errorTip.addClass("hide");
        });
        // 填充名字
        var nick = model.globalData.user.nick || '';
        that.find('.nickName').val(nick);

        //选择类型
        $(".choice-type").each(function (index, element) {
            var that_son = $(this);
            that_son.on("click", function () {
                $(".select").removeClass("isSelected");
                that_son.find('.select').addClass("isSelected");
                that.typeId = that_son.find('.select').attr('type');
            });
        });
        $(".choice-fault").each(function (index, element) {
            var that_son = $(this);
            that_son.on("click", function () {
                $(".fault").removeClass("isSelected");
                that_son.find('.fault').addClass("isSelected");
                that.fault = that_son.find('.fault').attr('fault');
            });
        });
        //屏幕类型下一步
        $(".inquiry-type-next").on("click", function () {
            $("#TypeSelect").addClass("hide");
            $("#ColorSelect").removeClass("hide");
            $("#LastSelect").addClass("hide");
            //that.updateInit(that.typeId);
        });
        //颜色下一步
        $(".inquiry-color-next").on("click", function () {
            $("#TypeSelect").addClass("hide");
            $("#ColorSelect").addClass("hide");
            $("#LastSelect").removeClass("hide");
        });
        //颜色上一步
        $(".inquiry-color-prv").on("click", function () {
            $("#TypeSelect").removeClass("hide");
            $("#ColorSelect").addClass("hide");
            $("#LastSelect").addClass("hide");
        });
        //last上一步
        $(".inquiry-last-prv").on("click", function () {
            $("#TypeSelect").addClass("hide");
            $("#ColorSelect").removeClass("hide");
            $("#LastSelect").addClass("hide");
        });
        this._watchEvent();
        this._els = this._getElements();
    }
    private _init() {
        $("#TypeSelect").removeClass("hide");
        $("#ColorSelect").addClass("hide");
        $("#LastSelect").addClass("hide");
    }
    //昵称校验
    private checkNick(info, dom, tip) {
        var reg = /^[\u4e00-\u9faf]+$/
        if (info == "") {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-nick').text(tip + '不能为空！')
            return false;
        } else if (!reg.test(info)) {
            $('.' + dom + 'Tip').removeClass("hide");
            $('.' + dom + 'Tip').find('.error-nick').text('请输入正确的' + tip);
            return false;
        } else {
            return true;
        }
    }
    public setCity(city) {
        this.find('.citySelect').text(city);
    }
    public setRepairId = function (repairId) {
        this.repairId = repairId;
    }
    //提交
    private _evt_repairSubmit() {
        var that = this;
        var mobile = model.globalData.user.phone;
        var cate = 2;//维修
        var sem = localStorage.getItem("repair_statistics");
        var nick = that.nick.val();
        var city = that.find('.citySelect').html();
        var prov = model.globalData.selected_province.province_name;
        var addr = that.addr.val();
        var prePostInfo = {
            mobile: mobile,
            cate: cate,
            prov: prov,
            city: city,
            sem: sem
        }
        if (that.checkNick(nick, 'nick', "称呼")) {
            api.PrePost(prePostInfo, function (data: any) {
                var repairId = data.id;
                var rapairInfo = {
                    id: repairId,
                    nickname: nick,
                    type: that.typeId,
                    prov: prov,
                    city: city,
                    addr: addr,
                    malf: that.fault,
                }
                api.ActivateDemand(rapairInfo, function (data: any) {
                    //发布成功，，直接打开聊天室（“已登录，无在建项目”，询价）
                    window.location.reload();
                    window.location.href = '/ChatRoom/?demand=' + data.demand_info.demand.id;
                }, function (error: tomato.PError) {
                    if (error.name == "401") {
                        //未登录
                        funs.tip(error.note, "", "fail");
                        setTimeout(function () {
                            window.location.reload();
                        }, 2000);
                    } else if (error.name == "41004002") {

                        //项目不存在
                        funs.tip(error.note, "", "fail");
                        setTimeout(function () {
                            window.location.reload();
                        }, 2000);
                    } else {
                        funs.tip(error.note, "", "fail");
                        setTimeout(function () {
                            window.location.reload();
                        }, 2000);
                    }
                });

            }, function (error: tomato.PError) {
                if (error.name == '401') {
                    //未登录，需弹出验证码弹窗进行登录
                    tomato.getVPresenter<code>(funs.moduleToUrl('m/user/Code'), (code) => {
                        code.setCallBack(mobile, function () { that._evt_repairSubmit(); });
                        tdom.open(code, tdom.DialogTarget.Blank, {
                            size: { width: '90%', height: tomato.DialogSize.Content },
                            masked: true,
                        });
                    });
                } else {
                    funs.tip(error.name, error.note, "fail");
                }
            });
        }
    }
    /**城市选择 */
    _evt_citySelect() {
        // tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/CitySelector'), (CitySelector) => {
        //     (CitySelector as any).setProv("Repair");
        //     funs.loadPage(CitySelector);
        // })
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/CitySelector'), (CitySelector) => {
            //登录窗
            (CitySelector as any).setProv("select");
            tdom.open(CitySelector, tdom.DialogTarget.Blank, {
                size: { width: '100%', height: "100%" },
                effect: tdom.DialogEffect.scale,
                bodyEffect: tdom.TurnEffect.cover,
                masked: true
            });
        });
    }
}
export = VPresenter;
