// views/m/project/PhoneInquiry/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as track from "views/global/common/Track";
import * as model from "views/global/common/Model";
import * as api from "views/global/common/API";
import * as Vue from "vue";
import demandDetail = require('views/m/DemandDetail');

class VPresenter extends project.VPresenter {
    private webim: any;
    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._watchEvent();
        this._els = this._getElements();
        var that = this;

        // var demandList = model.globalData.demand;
        var demandLists=model.globalData.demand;
        var demandNum;
        if(demandLists){
            demandNum=demandLists.length;
        }else{
            demandNum=0;
        }
        
        //如果没有进行中的项目
        if (demandNum == 0) {
            this.find('.no-demand').removeClass('hide');
            // 城市选择
            var loc = this.find('.location')[0];
            var selected_city = model.globalData.selected_city;
            new Vue({
                el: loc,
                data: {
                    city: selected_city
                },
            })

            /** 如果是晚上（19:00-－08:00），替换界面元素及事件 */
            var myDate = new Date();
            var hours = myDate.getHours();
            if (hours >= 21 || hours < 8) {
                this.find('.biginquiryBtn').removeAttr('href').html('<span class="icon-message02"></span>客服在线，立即咨询').attr('evt', 'openPop|setupConsult');
                // this.find('.bubbles').html('在线出报价哦…<em class="horn"></em>').css('width','1.1rem');
                this.find('.button').removeAttr('href').attr('evt', 'openPop|setupConsult');
            }
        } else {
            //有项目
            this.find('.has-demand').removeClass('hide');
            this.find('.demand-num').text(demandNum);
            //登陆且有项目的用户，数据绑定
            var pro_inquiry = this.find('.has-demand')[0];
            var demd = model.globalData.demand;
            var unread = model.globalData;
            // var demandLists = [{ name: "室内高清屏安装" }, { name: "户外高清屏维修" }, { name: "室内高清屏维修" }, { name: "户外高清屏安装" }];
            var pro: any = new Vue({
                el: pro_inquiry,
                data: {
                    demand: demd,
                    unread: unread,
                    demandLists: demandLists,
                },
            });
            if (!(demandLists.length == 1)) {
                //存在多个项目
                this.find('.show-more').removeClass('hide');
            }
        }

    }
    _evt_openCitySel() {
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/common/CitySelector'), (CitySelector) => {
            funs.loadPage(CitySelector);
        })
    }
    _evt_openPop() {
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/customService/serviceChatPane'), (serviceChatPane) => {
            tdom.open(serviceChatPane,'_blank',{size:{width:"100%",height:"100%"}});
        });
    }
    _evt_setup400(){
        track.setup_400();
        return true;
    }
    _evt_setupConsult(){
        track.setup_consult();
        return true;
    }
    /**展开项目列表 */
    _evt_showMore() {
        $('.has-demand').addClass("show-demand-list");
    }
    /**收起项目列表 */
    _evt_closeUp() {
        $('.has-demand').removeClass("show-demand-list");
    }
    /**打开详情 */
    _evt_openDetail(data, target, hit) {
        //修改选中类型
        var name = target.innerHTML;
        var index = $(target).attr("index");
        // alert(name + "::" + index);
        model.globalData.current_demand = model.globalData.demand[index];
        tomato.getVPresenter<demandDetail>(funs.moduleToUrl('m/DemandDetail'), (demandDetail) => {
            if(demandDetail.vpid.indexOf('demand=') >0){
                demandDetail.vpid = demandDetail.vpid.replace(/\?(demand=\d*)?/,'?demand='+model.globalData.current_demand.id);
            }else{
                demandDetail.vpid += '?demand='+model.globalData.current_demand.id;
            }
            funs.loadPage(demandDetail);
        })
        this._evt_closeUp();
    }
     /**打开聊天室 */
    _evt_openChatRoom() {
        //将临时数据更改为自己的项目的数据
        model.globalData.currIsSelfDemand = true;
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/ChatRoom'), (ChatRoom) => {
            //弹出详情页
            funs.loadPage(ChatRoom);
        })
        this._evt_closeUp();
    }
    /**报装新屏幕 */
    _evt_createDemand() {
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/project/InqTypeSelect'), (InqTypeSelect) => {
            //登录窗
            tdom.open(InqTypeSelect, tdom.DialogTarget.Blank, {
                size: { width: '90%', height: tomato.DialogSize.Content },
                effect: tdom.DialogEffect.scale,
                masked: true
            });
        });
    }
}
export = VPresenter;
