// views/m/Chating/index.ts
import * as tomato from "@po-to/tomato";
import * as tdom from "@po-to/tomato-jquery";
import * as project from "views/global/common/Project";
import * as funs from "views/global/common/Funs";
import * as api from "views/global/common/API";
import * as model from "views/global/common/Model";
import * as track from "views/global/common/Track";

class VPresenter extends project.VPresenter {
    private currGroup: any;//当前已经使用的问答组的index
    private changeRate: any;//随机变化的参数
    private data: any;// 对话数据
    private boxNum: any;// 方块个数
    private myTimes;//时间

    constructor(view: tomato.VPView, parent?: tomato.VPresenter, vpid?: string) {
        super(view, parent, vpid);
        this._watchEvent();
        this._els = this._getElements();
        var that = this;
        this.myTimes = new Date();
        this.currGroup = 9;//当前已经使用的问答组的index
        this.changeRate = 5; //随机变化的参数
        this.changeOnlineNum();
        // 对话数据
        if (this.view.parent().get(0).id == "Repair") {
            this.find('.text').html('一键发布报修，立即召集工程师与您洽谈，附近有<span class="engineerOnline">173</span>位工程师在线');
            this.data = [["屏幕移位置，怎么收费？", "要从哪里移到哪里？"], ["维修怎么收费？", "基础物料费加上人工费"], ["门头屏坏了能修吗？", "什么故障，有照片吗？"], ["广告屏部分不亮怎么回事？ ", "您能拍张照片来看看吗？"], ["屏幕点不亮了 ", "电源通电了吗？"], ["您好，字幕屏一半不显示怎么回事？", "用字幕编辑重新试试"], ["屏幕显示不全好修吗？", "能修"], ["明天你能过来修吗？ ", "可以，明天10点上门修"], ["能提供现场技术支持服务吗？", "可以的，什么时候需要？"], ["屏幕上有些地方不亮怎么回事？", "您先发个照片来看看故障情况"], ["屏幕重影了 ", "什么屏，门头屏还是广告屏？"], ["今天能上门修吗？", "能，我也在万县"], ["上门维修费用怎么算？", "维修费加路费"], ["能清洗屏幕吗？", "可以的，您是多大的屏幕？"], ["户外大屏有死点，好修吗？", "可以修"], ["屏幕清洗多少钱？", "您多大的屏幕，安装在哪？"], ["价格怎么算？", "得看更换多少部件，您那有备品吗？"], ["屏幕改字多少钱？", "远程指导您更改更便宜，上门加路费"], ["都在五角场，路费就免了罢 ", "本来也只收您物料费"], ["屏幕总是闪怎么回事？", "可能是控制卡问题"], ["安庆宜秀区能上门修吗？", "能，我就是安庆的"], ["我的屏幕半边不亮", "能发个故障图片过来吗？"], ["led屏下半截闪动是为什么", "需要现场诊断才知道"], ["洪山广场附近led屏维修", "我们全国都有服务网点的"], ["led显示屏显示不正常", "是颜色不对还是不显示？"], ["屏幕显示的图片变形了？", "能拍张图片过来看看吗？"], ["为什么led屏不亮？", "尝试重新插拔电源"], ["还有上门费？怎么算？", "按照距离算的"], ["如何让led显示屏不闪 ", "有可能是控制卡故障"], ["led显示屏怎么不亮了 ", "之前也出现过这种情况吗"], ["led显示屏老是闪 ", "检查一下电源"], ["是哪里出了问题？ ", "初步排除控制卡的问题"], ["屏幕内容怎么更改？ ", "内容是文字还是视频？"], ["led大屏维修多少钱？", "屏幕怎么了？"], ["怎么重启？ ", "您重新开一下电源"], ["每次改字都特别麻烦 ", "没关系，我们可以上门帮您改"], ["改一次字多少钱？", "可以先上门教您改，包教包会！"], ["好像是控制卡坏了 ", "什么型号的"], ["屏幕上有两三行不亮 ", "可能是灯珠坏了"], ["灯珠坏了要自己另买吗？", "您那还有备用的灯珠吗？"], ["买的时候好像有些备品 ", "您看看有多少灯珠"], ["屏幕进水了有死灯 ", "这个比较麻烦，得赶紧拆开"], ["我在萧山区，能尽快上们吗 ", "可以，萧山区哪里？"], ["您好，我的屏幕点不亮了 ", "电源都检查了吗"], ["电源没问题 ", "这个需要上门排查故障了"], ["黄石西塞山区能上门吗？", "我就在黄石，随时可以上门"], ["今天在家，您尽早来修 ", "好的，1点多就能到"], ["您这维修费用怎么算？ ", "得看您需要更换哪些部件"], ["有备品，有的部件不需要再买 ", "可以的，有备品更好"], ["我的是路升的屏，你们能修吗？", "控制卡是哪家的还记得吗？"], ["我去哪里看屏幕信息 ", "您找找合同，那上面都有"], ["诺瓦的控制卡 ", "好的，能把配置清单拍个照吗"], ["维修有承保吗？", "有的，180天重复故障免费重修"], ["好的，我1点钟在店里等您 ", "好的，到了与您联系"], ["您好，强力巨彩的屏能修吗？", "您好，户外还是室内的"], ["led显示屏怎么调多行字幕？", "你需要远程协助，还是上门服务"], ["您好，炫彩光电屏能修吗？", "可以的，什么故障呢？"], ["有几行不亮了 ", "这个能修，您想什么时候修呢？"], ["这是什么引起的不亮呢？", "肯能是线缆接触不良"], ["要是换个线缆得多少钱？", "这个按长度计算，不贵"], ["你们上门要额外收费吗？", "一般都是要按距离收费的"], ["我能不能买个线缆自己换？", "小心别弄坏其他地方就可以"], ["我尝试了更换了设置没用 ", "你的客户端版本号多少"], ["您怎么指导我操作 ", "你电话多少，我电话教你"], ["升级完成了，然后呢？ ", "您关机重启一下"], ["您好，楼体广告屏能修吗？", "在哪里"], ["太原尖草坪区今天能上门吗 ", "我在迎泽区，上门很方便"], ["广告屏维修多少钱？", "您什么故障？"], ["色彩显示不对 ", "能发个照片吗"], ["矫正一次色彩能保持多久 ", "具体要看您是什么屏幕"], ["一年可以保持吗？", "质量好的屏幕都是半年一维护的"], ["可以把单色的屏幕换成双色吗？", "不可以的"], ["明天可以上门吗？", "随时都可以"], ["您好，在吗 ", "在的，有什么要咨询吗"], ["维修价能不能便宜点？", "路费都给您免了，亲"], ["户外的", "您的屏幕是户外还是室内屏？"], ["三彩光电的", "哪家的屏幕？"], ["总是会跳，一闪一闪的", "屏幕什么故障，能描述一下吗？"], ["字幕屏", "是广告屏还是字幕屏？"], ["好修吗？", "可能是排线坏了 "], ["拍照给您看看", "备品备料有哪些？"], ["后凯光电的屏幕", "什么品牌的屏幕？"], ["才用了1年多", "用了多久呢？"], ["绵阳", "你在四川哪里？"], ["稍等", "把您的客户端界面拍个照看下 "], ["去哪里升级？", "你的版本太低了，需要升级 "], ["西安脸谱广场", "您在哪里？"], ["哦，那矫正一次多少钱呢？", "这需要色彩矫正 "], ["好，等一下", "能发个故障图片过来吗？  "], ["屏幕闪动", "led屏出现什么故障？ "], ["对有色差", "颜色异常是吗？ "], ["利亚德的", "您的屏幕是什么牌子？ "], ["换灯珠怎么计费？", "出现死点，需要更换灯珠"], ["士兰明芯的，型号忘记了", "你的控制卡型号是什么？ "], ["不能播放视频", "你的屏幕哪里坏了？"], ["插了，还是不亮", "插电源了吗？"], ["半个屏吧", "不亮的面积大概有多大？"], ["有死点", "屏幕出了什么故障？ "], ["有的，两个模组", "你那有备品备料吗？ "], ["今年3月份就到期了", "质保期到什么时候？ "], ["郑州的能修吗", "专业维修led屏，欢迎联系 "], ["您好，我的屏幕不亮了", "您好我是工程部，林生  "], ["对，还有死点", "屏幕是出现黑线吗？"], ["还没，厂家收费比较高", "您的屏幕保质期过了吗？"], ["是呀，给我的维修报价很贵", "厂家卖的多，售后也很忙 "], ["控江路", "您在杨浦哪里？"], ["有家报价比你便宜", "您好，昨天的报价能接受吗 "], ["畅捷的", "他们跟你换什么品牌控制卡 "], ["我查了感觉差不多", "畅捷的没有中航的好 "], ["便宜多少", "畅捷的我也有，确实便宜些 "], ["是的", "你好，门头屏维修是吗？"], ["对的，能修吗", "您是全彩的门头屏？"], ["屏幕有黑点", "什么故障呢？"], ["您稍等", "多大面积的黑点，有照片吗？"], ["哦，那是什么问题呢？", "你这个屏幕还不算严重 "], ["能修吗", "可能是零部件问题 "], ["好换吗，我家有备用灯珠", "换掉灯珠可能靠谱 "], ["维修费用怎么算呢", "有备品更方便 "], ["上门费怎么算呢", "只要上门费就好 "], ["桃江县", "您在岳阳哪里？ "], ["这么算有点贵吧", "按车程算，我在资阳区 "], ["好吧，能尽快上门不", "别家也是这么计费的 "], ["那就今天下午吧", "可以的，随时可以上门 "], ["我去现拍一个，稍等", "您好，能上传故障照片吗？"], ["是的，看的人眼睛都花了", "屏幕花屏是吗？"], ["3、4天了", "故障多久了？"], ["强力的", "什么品牌的屏幕 "], ["快两年的吧", "使用多久了？"], ["好是好，就是坏起来挺烦人", "那您这个还算不错了 "], ["多少钱能修好", "您这个还不算难修 "], ["好的", "我跟您电话聊 "], ["你好", "您好，led显示屏维修是吗 "], ["屏幕出现方格", "您的屏幕具体什么故障呢？ "], ["可以，我找一下", "方便先拍个照片看看吗 "], ["今天坏的", "坏了多久？"], ["好的，我去看看", "您先检查线头有没有插好 "], ["没，正准备问您呢", "您好，您的屏幕好了没？"], ["指示灯都没亮", "控制主机能打开吗？"], ["好修吗", "应该是控制主机部件坏了 "], ["多少钱呢", "得上门看，可能要换部件 "], ["整屏画面晃动", "您好，屏幕什么故障呢？"], ["看了，正常的", "电脑与屏幕的线连接正常吗？"], ["您这边能修吗", "可能是发送卡故障了？"], ["不记得了", "您记得是哪家的发送卡吗？"], ["飞控的", "合同上有，你去看看 "], ["两年多了", "您的屏幕使用几年了？"], ["要自己买吗", "可能需要跟换发送卡了 "], ["我拍个图给你", "你的发送卡什么型号 "]];
        } else {
            this.data = [["洲明这个牌子怎么样？", "洲明小间距品质很不错的"], ["南昌西湖区能来装吗？", "能，我就在西湖区，1小时内上门"], ["想在饭店里装块大屏", "饭店多大面积呢？"], ["我想安个8平米的广告屏", "是户外全彩的吗？"], ["想安P4户外全彩的屏", "具体尺寸是多少？"], ["公路边上的高架屏能装吗？", "可以的"], ["价格怎么算呢？", "设备材料费和路费"], ["明天你能过来安装吗？", "可以，我们同城"], ["广州led屏幕p6多少钱？", "你想多大面积的屏幕"], ["p4是什么意思？", "是间距，数字越小，画面越清晰"], ["p4和p6显示区别大吗？", "p4更清晰，p6户外使用"], ["您这价格能再便宜些吗？", "已经免安装费了，亲"], ["送货上门吗？有安装费吗？", "送货上门，安装免费"], ["你家怎么比别人贵？", "蓝普是大品牌，质量过硬"], ["支持跨省安装吗？", "支持，我们全国业务都能做"], ["小镇上装门头屏能上门吗？", "可以寄货到家，自己安装更划算哦"], ["led屏一平方多少钱？", "对颜色间距品牌有什么要求"], ["p3led全彩显示屏价格", "小间距的会稍微贵一点"], ["p4led全彩显示屏怎么卖", "不同品牌价格也不同"], ["p5led全彩屏多钱？", "您一次要买多少平米？"], ["p10户外全彩显示屏多少钱？", "1200一平安装"], ["led电子屏什么牌子好", "你的屏幕是用来做什么的？"], ["led显示屏有哪些品牌", "很多，要看屏幕具体需求"], ["led显示屏厂家有哪些", "厂家对起买量是有限制的"], ["我要防水的户外显示屏", "户外led屏都防水的"], ["自己安装？我不会呀！", "我们会远程指导您安装"], ["小间距弧形的，能做吗？", "弧形是需要定制的"], ["小间距有什么好处？", "间距越小，清晰度越高"], ["找你安装，质保多久？", "行业内都是质保1年的"], ["洲明led小间距，怎么卖？", "您具体要哪种间距？"], ["户外全彩屏，播广告用", "那P8的就可以了"], ["室内会议室的led屏", "建议您安装P3或P4的"], ["明天可以上门安装吗", "您备品齐全，我随时可上门"], ["找北京led屏供应商", "具体安装在北京哪里？"], ["面积大概12平米", "是户外显示广告用吗"], ["主要用来播广告", "建议选择户外全彩P8屏"], ["led屏里面带音响吗？", "可以带音响"], ["想问下户外广告屏多少钱？", "要什么规格的"], ["能做P3的全彩led屏吗", "可以，我们专业做P3的"], ["你有洲明的小间距led屏吗？", "有的，你要哪个间距"], ["蓝普这个牌子怎么样", "蓝谱性价比很高"], ["强力质量怎么样", "强力户外屏不错"], ["户外用哪个间距比较好？", "P6或者P8都可以"], ["室内用p3可以吗？", "您的观看距离有多远"], ["全彩p5怎么收费？", "不同品牌价格也不同的"], ["您报价多少 ？", "3500全包"], ["上门费用怎么收？", "按照距离远近计算"], ["2平米的门头，大概多少钱？", "大概2000块以内"], ["配电柜需要自己提供吗？", "我们可以提供的"], ["好几家的报价都比你低", " 亲，是同样的配置吗？"], ["你们是led屏厂家吗？", "我们的货源是厂家直供的"], ["什么时候提供具体的报价？", "具体报价我电话跟您说"], ["质保多久？", "质保您2年，没问题"], ["你们提供配件吗？", "如果需要，是可以提供的"], ["点间距是什么？", "灯珠距离，间距越小越清晰"], ["比较急，今天能上门吗？", "可以的，我就在浉河区"], ["您好，在吗?", "在，屏幕要求已在详情里说明了"], ["可以邮寄过来吗？", "可以，可包邮"], ["为什么需要钢结构？", "是用来固定用的"], ["能提供发票吗？", "可以邮寄发票给您"], ["led大屏幕多少钱？", "这要看具体的规格要求"], ["配置都差不多", "给您做广告屏，我们用蓝普的屏"], ["以后修改文字能帮我改吗？", "必须可以呀，上门改都没问题"], ["不着急，我就是先问问", "没关系，您还有哪些想要了解的"], ["人工费怎么算？", "一天一人300块"], ["大概需要几天安完？", "这个比较复杂，需要3天左右"], ["工期大概多久？", "你这个项目有些复杂"], ["支持播放外国语言吗？", "我们支持十六国语言显示"], ["可以连接WiFi吗", "可以的，里面有WiFi模块"], ["你们售后是怎样的？", "我们质保您1年"], ["那把合同给我吧", "好的，明天之前发您"], ["什么时候可以上门来看一下", "今天下午就可以"], ["可以监控屏幕播放的内容吗？", "需要额外接摄像头"], ["散热设备谁提供？", "如果需要，我们可以提供"], ["高速公路旁边的那种，能做吗？", "可以的，最好有场地照片"], ["深圳市南山区", "您想安在哪里？"], ["平桥区万象城", "您的广告屏在哪里使用"], ["户外广告屏", "您好，需要什么屏？"], ["你好，在北京海淀区", "你好，什么规格？安装在哪？"], ["大概2平米", "专业做门头，你要多大面积"], ["p4安在会议室", "什么规格，安装在哪？"], ["好的，谢谢", "在吗？电话联系给你报价"], ["是的，美尚百货的楼体上", "是装在楼体上的吗？"], ["是的，在马路边上", "你这是要装个高炮类型的？"], ["6月8号就要用了", "什么时候使用？"], ["全彩的价格怎么样呢？", "门头屏用全彩的更醒目哦！"], ["那您给装p几的呢？", "才5㎡，做全彩也就2000"], ["我跟隔壁店一起装能再优惠不", "p10的就可以了，比较实惠！"], ["好的，谢谢", "有什么问题随时联系我"], ["不会安装", "门头屏您可以自己安装，更便宜"], ["p5的", "你的屏幕间距是多大？"], ["太麻烦了，还是你们上门装吧", "邮寄给您自己装，有视频教程"], ["那太好了，感谢", "我也在经开区，安装费就不收您了"], ["嗯，已经发你图片了", "有图片更好报价"], ["大概15米宽、8米高", "您装个多大的，长宽多少？"], ["上海市静安区", "屏幕在什么位置？"], ["就是啊，所以才换成led屏", "咱这冬天雾霾重，得装个亮度高的"], ["可是价格还是有点贵", "强力也是大牌子，不容易坏"], ["做广告屏我不能只图便宜", "洲明虽然质量好，但是价格也贵呀"], ["10米多的样子", "您的屏幕需要架多高？"], ["需要", "散热设备需要我们提供吗？"], ["浦东张江", "你在上海哪儿？"], ["好的", "广州地区，可以上门"], ["上门费怎么计算？", "您好，上门费另算的"], ["清晰一点", "您对清晰度有什么要求？"], ["那我选择p8吧", "p8比P10要更清晰一些"], ["不需要，户外信息屏那种", "需要高空安装吗？"], ["3×5米的样子", "大概需要多少平米？"], ["质量要好，价格好商量", "你对品牌有要求吗？"], ["室内，会议室用的", "室内还是室外的屏幕？"], ["2米吧", "最佳视距有多远？"], ["播放广告视频", "led屏是用来干什么的？"], ["哪有什么可以替换吗", "你的这个模组已经停产了"], ["价格怎么样", "洲明是高端品牌，质量好"], ["它有小间距的屏吗", "蓝普性价比很高的"], ["全彩，也要播图片的", "全彩还是双色？"], ["后天上午可以吗", "什么时候方便，我们上门安装"], ["10万以内", "有大概的预算吗？"], ["有质保吗？", "一平米大概1200元左右"], ["最好可以上门帮调试", "我们可以上门帮忙安装"], ["我在朝阳区", "本市30分钟就可以上门"], ["主要是图片和文字", "屏幕用来播放什么？"], ["那长沙可以上门吗", "我们全国都有网点的"], ["您打我电话就行", "上门安装led屏，欢迎联系"], ["有p25的吗？", "专业制作小间距"], ["是的，能做吗？", "你是要做弧形屏是吗？"], ["需要，在室外", "需要防水吗？"], ["不好意思，刚刚没听到", "怎么打你电话打不通啊？"], ["你们报价多少？", "你这个预算，太低了"], ["都是什么价格呢？", "餐厅内使用，用p4或p5比较好！"], ["是的，承重墙", "led屏挂的墙是实心墙吗？"], ["好的，我知道了", "你这个需要额外的钢结构"], ["那就p3吧", "P3比P4更清晰，价格差不多"], ["价格贵吗？", "现在户外都用P8了"], ["4×6米的户外屏", "屏幕尺寸多大？"], ["8万，不能再多了", "您的预算大概是多少？"], ["对，25层的外面", "屏幕是安在大楼外面吗？"], ["室内两块屏，室外一块屏", "您一共要安几块屏？"], ["医院里面展示信息的", "室内屏是用来做什么的？"], ["有人，昨天来的", "您好，屏幕有人去安装吗"], ["通过平台联系我", "我怎么联系您？"], ["宝安区海雅缤纷城", "具体安屏位置是哪里？"], ["不能再便宜了吗？", "最低报价12万"], ["什么时候可以上门？", "这个价格我可以做"], ["没有，你有推荐吗？", "你有指定的品牌吗？"], ["p8多少钱？", "p10单色没货了，p8可以吗？"], ["p10怎么卖", "p12以上的已经停产，p10最实惠！"], ["30平米的样子", "安装屏幕的房间有多大？"], ["费电吗？", "户外led屏颜色都比较亮"], ["全彩多少钱一平？", "门头屏现在用全彩效果更好"]];
        }

        this.initBoxes();
        $(window).resize(function () {
            that.initBoxes();
        });
        /** 不同情形不同时间段，“开始洽谈”按钮功能变化 */
        if (model.globalData.demand.length == 0) {
            // 如果是晚上（21:00-－08:00），替换按钮元素的点击事件 
            var myDate = new Date();
            var hours = myDate.getHours();
            if (hours >= 21 || hours < 8) {
                this.find('.btn').removeAttr('href').attr('evt', 'openPop');
            }
        } else {
            this.find('.btn').removeAttr('href').attr('evt', 'openChatRoom');
        }

    }
    initBoxes() {
        var scrWid = $(window).width();
        // 三排三列显示,改变变化速率
        if (scrWid <= 768) {
            this.currGroup = 8;
            this.changeRate = 2;
        } else if (scrWid <= 992) {
            this.currGroup = 9;
            this.changeRate = 2;
            // 四排四列显示,改变变化速率
        } else if (scrWid <= 1200) {
            this.currGroup = 16;
            this.changeRate = 3;

            // 五排五列显示,改变变化速率
        } else if (scrWid > 1200) {
            this.currGroup = 25;
            this.changeRate = 5;
        }
        // 取前n条插入dom
        var tpl = '';
        for (var i = 0; i < this.currGroup; i++) {
            tpl += '<div class="box lg-5 md-4 sm-3 xs-2"><div class="box-y"><p class="text">' + this.data[i][0] + '</p><span class="icon-use"></span></div><div class="box-b"><span class="icon-supplier"></span><p class="text">' + this.data[i][1] + '</p></div></div>'
        }
        this.find('.tpl').html(tpl);
        this.randomBox();
        this.boxNum = this.currGroup;
    }
    // 重复随机选取多个方块，执行变化方法
    randomBox() {
        var that = this;
        setInterval(function () {
            var i = 0;
            do {
                var random = Math.round(Math.random() * that.changeRate);
                if (2 * i + random > that.boxNum) {
                    return
                }
                var box = that.find('.box').eq(2 * i + random);
                that.changeStatus(box);
                i += random;
            } while (i < that.boxNum);
        }, 3000);
    }
    // 单个方块的变化
    changeStatus(jqDom) {
        // 累加要取得的词条索引值  (-3是因为循环调用此方法会出现currGroup的值溢出的情况导致报错，无奈只好如此)
        if (this.currGroup >= this.data.length - 3) {
            this.currGroup = 0;
            return
        } else {
            this.currGroup++;
        }
        // 生成随机结果备用
        var flag1, flag2, flag3, flag4, flag5, random1 = Math.random(), random2 = Math.random();
        var left, right, top, bottom;
        if (random2 > 0.5) {
            flag2 = 'right';
            flag3 = 'left';
        } else {
            flag2 = 'left';
            flag3 = 'right';
        }
        if (Math.round(random2 * 10) % 2 == 0) {
            flag4 = 'top';
            flag5 = 'bottom';
        } else {
            flag4 = 'bottom';
            flag5 = 'top';
        }
        //设置每一个动画的随机延迟时间
        var randomDelay = Math.round(random1 * 1000);
        // 判断当前状态，针对性改变
        var b = jqDom.find('.box-b');
        var y = jqDom.find('.box-y');
        if (!this.data[this.currGroup]) {
            return
        }
        var text_b = this.data[this.currGroup][1];
        var text_y = this.data[this.currGroup][0];
        var textb = b.find('.text');
        var texty = y.find('.text');
        if (b.css('height') == '0px') {
            //蓝色显示
            b.delay(randomDelay).css('height', '100%');
            textb.text(text_b).delay(300).css('width', '62%');
            // 黄色块
            y.delay(randomDelay).find('.text').css('width', '0px').text(text_y);
        } else if (b.css('width') == '0px') {
            //蓝色显示
            b.delay(randomDelay).css('width', '100%');
            textb.text(text_b).delay(300).css('width', '62%');
            // 黄色块
            y.delay(randomDelay).find('.text').css('width', '0px').text(text_y);
        } else {
            /** 随机改变色块的动画方向 */
            random1 > 0.5 ? flag1 = 'height' : flag1 = 'width';
            b.delay(randomDelay).css(flag1, '0px').delay(800).find('.text').css('width', '0px');
            y.delay(randomDelay).find('.text').text(text_y).delay(300).css('width', '62%');
            setTimeout(function () {
                jqDom.find('.box-b').css({ flag2: 'auto', flag3: '0', flag4: 'auto', flag5: '0' });
                textb.text(text_b);
            }, 1000);
        }
    }
    /**分时间段动态显示工程商数量 */
    changeOnlineNum() {
        var customerNum = 2500;
        var coefficient = 3.73;
        this.myTimes = new Date();
        var hours = this.myTimes.getHours();
        var num1, num2;
        if (hours >= 24 || hours < 1) {
            num1 = Math.round(customerNum * 0.02);
            num2 = Math.round(Math.round(customerNum * 0.02) * 3.73);
        } else if (hours >= 1 && hours < 2) {
            num1 = Math.round(customerNum * 0.012);
            num2 = Math.round(Math.round(customerNum * 0.012) * 3.73);
        } else if (hours >= 2 && hours < 3) {
            num1 = Math.round(customerNum * 0.01);
            num2 = Math.round(Math.round(customerNum * 0.01) * 3.73);
        } else if (hours >= 3 && hours < 4) {
            num1 = Math.round(customerNum * 0.013);
            num2 = Math.round(Math.round(customerNum * 0.013) * 3.73);
        } else if (hours >= 4 && hours < 5) {
            num1 = Math.round(customerNum * 0.015);
            num2 = Math.round(Math.round(customerNum * 0.015) * 3.73);
        } else if (hours >= 5 && hours < 6) {
            num1 = Math.round(customerNum * 0.02);
            num2 = Math.round(Math.round(customerNum * 0.02) * 3.73);
        } else if (hours >= 6 && hours < 7) {
            num1 = Math.round(customerNum * 0.015);
            num2 = Math.round(Math.round(customerNum * 0.015) * 3.73);
        } else if (hours >= 7 && hours < 8) {
            num1 = Math.round(customerNum * 0.02);
            num2 = Math.round(Math.round(customerNum * 0.02) * 3.73);
        } else if (hours >= 8 && hours < 9) {
            num1 = Math.round(customerNum * 0.023);
            num2 = Math.round(Math.round(customerNum * 0.023) * 3.73);
        } else if (hours >= 9 && hours < 10) {
            num1 = Math.round(customerNum * 0.021);
            num2 = Math.round(Math.round(customerNum * 0.021) * 3.73);
        } else if (hours >= 10 && hours < 11) {
            num1 = Math.round(customerNum * 0.032);
            num2 = Math.round(Math.round(customerNum * 0.032) * 3.73);
        } else if (hours >= 11 && hours < 12) {
            num1 = Math.round(customerNum * 0.022);
            num2 = Math.round(Math.round(customerNum * 0.022) * 3.73);
        } else if (hours >= 12 && hours < 13) {
            num1 = Math.round(customerNum * 0.019);
            num2 = Math.round(Math.round(customerNum * 0.019) * 3.73);
        } else if (hours >= 13 && hours < 14) {
            num1 = Math.round(customerNum * 0.029);
            num2 = Math.round(Math.round(customerNum * 0.029) * 3.73);
        } else if (hours >= 14 && hours < 15) {
            num1 = Math.round(customerNum * 0.014);
            num2 = Math.round(Math.round(customerNum * 0.014) * 3.73);
        } else if (hours >= 15 && hours < 16) {
            num1 = Math.round(customerNum * 0.013);
            num2 = Math.round(Math.round(customerNum * 0.013) * 3.73);
        } else if (hours >= 16 && hours < 17) {
            num1 = Math.round(customerNum * 0.025);
            num2 = Math.round(Math.round(customerNum * 0.025) * 3.73);
        } else if (hours >= 17 && hours < 18) {
            num1 = Math.round(customerNum * 0.018);
            num2 = Math.round(Math.round(customerNum * 0.018) * 3.73);
        } else if (hours >= 18 && hours < 19) {
            num1 = Math.round(customerNum * 0.019);
            num2 = Math.round(Math.round(customerNum * 0.019) * 3.73);
        } else if (hours >= 19 && hours < 20) {
            num1 = Math.round(customerNum * 0.026);
            num2 = Math.round(Math.round(customerNum * 0.026) * 3.73);
        } else if (hours >= 20 && hours < 21) {
            num1 = Math.round(customerNum * 0.03);
            num2 = Math.round(Math.round(customerNum * 0.03) * 3.73);
        } else if (hours >= 21 && hours < 22) {
            num1 = Math.round(customerNum * 0.02);
            num2 = Math.round(Math.round(customerNum * 0.02) * 3.73);
        } else if (hours >= 22 && hours < 23) {
            num1 = Math.round(customerNum * 0.02);
            num2 = Math.round(Math.round(customerNum * 0.02) * 3.73);
        } else if (hours >= 23 && hours < 24) {
            num1 = Math.round(customerNum * 0.01);
            num2 = Math.round(Math.round(customerNum * 0.01) * 3.73);
        }
        if (this.view.parent().get(0).id != "Repair") {
            num1 += 1;
            num2 += 2;
        }
        this.find('.customerOnline').text(num1);
        this.find('.engineerOnline').text(num2);
    }
    // 点击开始洽谈事件
    _evt_openPop() {
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/customService/serviceChatPane'), (serviceChatPane) => {
            tdom.open(serviceChatPane, '_blank', { size: { width: "100%", height: "100%" } });
        });
    }
    // 点击开始洽谈事件
    _evt_openChatRoom() {
        tomato.getVPresenter<tomato.VPresenter>(funs.moduleToUrl('m/ChatRoom'), (ChatRoom) => {
            funs.loadPage(ChatRoom);
        });
    }
    _evt_track() {
        /** 不同情形不同时间段，“开始洽谈”按钮功能变化,统计也变化 */
        if (model.globalData.demand.length == 0) {
            // 如果是晚上（21:00-－08:00），替换按钮元素的点击事件 
            var myDate = new Date();
            var hours = myDate.getHours();
            if (this.view.parent().get(0).id == "Repair") {
                if (hours >= 21 || hours < 8) {
                    track.repair_consult();
                } else {
                    track.repair_400();
                }
            } else {
                if (hours >= 21 || hours < 8) {
                    track.setup_consult();
                } else {
                    track.setup_400();
                }
            }
        }
        return true;
    }
}
export = VPresenter;
